### PR TITLE
Refactor utility hook

### PR DESCRIPTION
**Pull request by onderkalaci**
We want to be able to find the "top-level" DDL commands
(not internal/cascading ones). To achieve that, we have
some refactoring.

### CODE REVIEW

File: src/backend/distributed/commands/utility_hook.c [hunk 1]
```diff
@@ -145,6 +157,108 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+/*
+ * Citus_ProcessUtility is a helper function for multi_ProcessUtility where majority
+ * of the Citus specific utility statements are handled here. The distinction between
+ * both functions is that Citus_ProcessUtility does not handle CALL and DO statements.
+ * The reason for the distinction is implemented to be able to find the "top-level" DDL
+ * commands (not internal/cascading ones). UtilityHookLevel variable is used to achieve
+ * this goal.
+ */
+static void
+Citus_ProcessUtility(PlannedStmt *pstmt,
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:252]: What about `multi_ProcessUtility_Internal`  ?
Or what about using `Citus_ProcessUtility` for above function and naming this as `Citus_ProcessUtility_Internal` ?

File: src/backend/distributed/commands/utility_hook.c [hunk 2]
```diff
@@ -73,8 +73,21 @@ static bool shouldInvalidateForeignKeyGraph = false;
 static int activeAlterTables = 0;
 static int activeDropSchemaOrDBs = 0;
 
+/*
+ * Number of Utility hook levels we are currently in, excluding
+ * DO blocks and Stored Procedures.
+ */
+int UtilityHookLevel = 0;
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:80]: should we add this in other pr ?

File: src/backend/distributed/commands/utility_hook.c [hunk 3]
```diff
@@ -84,13 +97,13 @@ static bool IsDropSchemaOrDB(Node *parsetree);
 
 
 /*
- * CitusProcessUtility is a convenience method to create a PlannedStmt out of pieces of a
+ * ProcessUtilityForNode is a convenience method to create a PlannedStmt out of pieces of a
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:100]: maybe `ProcessUtilityCommandForNode`

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:100]: (also `ForNode` part sounds a bit confusing to me)

File: src/backend/distributed/commands/utility_hook.c (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:252]: I think I could use `multi_ProcessUtilityInternal `

File: src/backend/distributed/commands/utility_hook.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:100]: Postgres doesn't seem to have `command` terminology for this purpose?

File: src/backend/distributed/utils/role.c [hunk 4]
```diff
@@ -96,11 +96,11 @@ worker_create_or_alter_role(PG_FUNCTION_ARGS)
...
-		CitusProcessUtility(parseTree,
-							createRoleUtilityQuery,
-							PROCESS_UTILITY_TOPLEVEL,
-							NULL,
-							None_Receiver, NULL);
+		ProcessUtilityForNode(parseTree,
+							  createRoleUtilityQuery,
+							  PROCESS_UTILITY_TOPLEVEL,
+							  NULL,
+							  None_Receiver, NULL);
```
Comment (onurctirtir) [inline src/backend/distributed/utils/role.c:103]: probably not related with this pr but I'm wondering why codecov claims that this is not covered

File: src/backend/distributed/commands/utility_hook.c [hunk 5]
```diff
@@ -145,6 +151,96 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+							 ProcessUtilityContext context,
+							 ParamListInfo params,
+							 struct QueryEnvironment *queryEnv,
+							 DestReceiver *dest,
+							 QueryCompletionCompat *completionTag)
+{
+	Node *parsetree = pstmt->utilityStmt;
+	List *ddlJobs = NIL;
 
 	bool isCreateAlterExtensionUpdateCitusStmt = IsCreateAlterExtensionUpdateCitusStmt(
```
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:245]: probably need to move lines [245:262] to `multi_ProcessUtility`

```c
bool isCreateAlterExtensionUpdateCitusStmt = IsCreateAlterExtensionUpdateCitusStmt(
...
...
if (!CitusHasBeenLoaded())
```

File: src/backend/distributed/commands/utility_hook.c [hunk 6]
```diff
@@ -145,6 +151,96 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+/*
+ * multi_ProcessUtilityInternal is a helper function for multi_ProcessUtility where majority
+ * of the Citus specific utility statements are handled here. The distinction between
+ * both functions is that Citus_ProcessUtility does not handle CALL and DO statements.
+ * The reason for the distinction is implemented to be able to find the "top-level" DDL
+ * commands (not internal/cascading ones). UtilityHookLevel variable is used to achieve
+ * this goal.
+ */
+static void
+multi_ProcessUtilityInternal(PlannedStmt *pstmt,
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:234]: Would it be a stretch to just call this ProcessDDLCommand?

I know Explain is in there as well, but it seems close enough.

File: src/backend/distributed/commands/utility_hook.c [hunk 7]
```diff
@@ -145,6 +151,96 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
+
+		return;
+	}
+	else if (IsA(parsetree, DoStmt))
+	{
+		/*
+		 * All statements in a DO block are executed in a single transaciton,
+		 * so we need to keep track of whether we are inside a DO block.
+		 */
+		DoBlockLevel += 1;
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:201]: not important right now, but it seems this basically does the same thing as StoredProcedureLevel.

File: src/backend/distributed/commands/alter_table.c [hunk 8]
```diff
@@ -621,8 +621,8 @@ ConvertTable(TableConversionState *con)
 		Node *parseTree = ParseTreeNode(tableCreationSql);
 
 		RelayEventExtendNames(parseTree, con->schemaName, con->hashOfName);
-		CitusProcessUtility(parseTree, tableCreationSql, PROCESS_UTILITY_TOPLEVEL,
-							NULL, None_Receiver, NULL);
+		ProcessUtilityForNode(parseTree, tableCreationSql, PROCESS_UTILITY_TOPLEVEL,
```
Comment (marcocitus) [inline src/backend/distributed/commands/alter_table.c:624]: What does "Node" refer to?

it makes me thing of a worker node.

File: src/backend/distributed/commands/utility_hook.c (hunk 6 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/utility_hook.c:254]: There is also COPY - VACUUM etc. as well, so DDL might be a little stretch. 

Some things that I considered:

- SingleStatementProcessUtility
- SingleCommandProcessUtility


But those also doesn't fit well

Comment (marcocitus) [reply src/backend/distributed/commands/utility_hook.c:254]: Maybe just ProcessUtilityInternal then, the multi_ prefix adds very little

**Review by marcocitus — APPROVED** [review]

**Review by onurctirtir — APPROVED** [review]

**Pull request merged**